In [ ]:
from datetime import datetime

from PIL import Image
import matplotlib.pyplot as plt
import matplotlib.patches as patches
from tqdm import tqdm
import os
import json
import pandas as pd


import matplotlib as mlp
import matplotlib.font_manager as fm
font_name = fm.FontProperties(fname = "C:/Windows/Fonts/malgun.ttf").get_name()
plt.rc("font", family = font_name)
mlp.rcParams["axes.unicode_minus"] = False

In [ ]:
# # 모든 열과 행을 출력할 수 있도록 설정
# pd.set_option("display.max_columns", None)
# pd.set_option("display.max_rows", None)

In [ ]:
# # 기본 설정으로 되돌리기
# pd.reset_option("display.max_columns")
# pd.reset_option("display.max_rows")

## 시설 토마토 train, val split

----------------

# Train 정상

In [30]:
# laptop
# bb_dir = "C:/juna27/PythonWork/AI/data/tomato_set/training/[라벨]11.토마토_0.정상"

In [42]:
# bb_dir = "C:/juna27/PythonWork/AI/data/tomato_label/tomato_set/training/[라벨]11.토마토_0.정상"

# Train 질병

In [43]:
# laptop
# bb_dir = "C:/juna27/PythonWork/AI/data/tomato_set/training/[라벨]11.토마토_1.질병"

In [84]:
# bb_dir = "C:/juna27/PythonWork/AI/data/tomato_label/tomato_set/training/[라벨]11.토마토_1.질병"

-----------

# Val 정상

In [64]:
# laptop
# bb_dir = "C:/juna27/PythonWork/AI/data/tomato_set/validation/[라벨]11.토마토_0.정상"

In [103]:
# bb_dir = "C:/juna27/PythonWork/AI/data/tomato_label/tomato_set/validation/[라벨]11.토마토_0.정상"

# Val 질병

In [87]:
# laptop
# bb_dir = "C:/juna27/PythonWork/AI/data/tomato_set/validation/[라벨]11.토마토_1.질병"

In [104]:
# bb_dir = "C:/juna27/PythonWork/AI/data/tomato_label/tomato_set/validation/[라벨]11.토마토_1.질병"

------------------

## Train 증강(질병 2종)

In [29]:
# bb_dir = "C:/juna27/PythonWork/AI/data/tomato_label/tomato_set/training/[라벨]11.토마토_9.증강"

## Val 증강(질병 2종)

In [106]:
# bb_dir = "C:/juna27/PythonWork/AI/data/tomato_label/tomato_set/validation/[라벨]11.토마토_9.증강"

------------------

## 라벨링 데이터 확인

In [3]:
label_names = os.listdir(bb_dir)
len(label_names),label_names

(19030,
 ['V006_77_1_18_11_03_13_1_3248b_20201005_57_a0000.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201005_57_a0001.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201005_57_a0002.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201005_57_a0003.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201005_57_a0004.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201005_57_a0005.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201005_57_a0006.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201005_57_a0007.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201005_57_a0008.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201005_57_a0009.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201012_60_a0000.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201012_60_a0001.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201012_60_a0002.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201012_60_a0003.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201012_60_a0004.jpg.json',
  'V006_77_1_18_11_03_13_1_3248b_20201012_60_a0005.jpg.json',


In [4]:
label_names[0]

'V006_77_1_18_11_03_13_1_3248b_20201005_57_a0000.jpg.json'

In [5]:
# 경로 확인 !!!! 위에서 잘 선언 했는지 꼭 확인

# bb_dir = "위에서 선언하는 것으로 수정, 실수 하지 않게 메모 !"

# load
bb_json_data = []
for file_name in tqdm(os.listdir(bb_dir)):
    if file_name.endswith(".json"):
        file_path = os.path.join(bb_dir, file_name)
        with open(file_path, "r", encoding="utf-8") as target_file:
            bb_json_data.append(json.load(target_file))

100%|████████████████████████████████████████████████████████████████████████████| 19030/19030 [03:11<00:00, 99.27it/s]


In [6]:
json_data = bb_json_data.copy()

In [7]:
json_data[0]

{'description': {'image': 'V006_77_1_18_11_03_13_1_3248b_20201005_57_a0000.jpg',
  'date': '2020/10/05',
  'worker': '',
  'height': 3024,
  'width': 4032,
  'task': 77,
  'type': 1,
  'region': 4,
  'original': 'V006_77_1_18_11_03_13_1_3248b_20201005_57.jpg'},
 'annotations': {'disease': 18,
  'crop': 11,
  'area': 3,
  'grow': 13,
  'risk': 1,
  'points': [{'xtl': 908, 'ytl': 712, 'xbr': 3104, 'ybr': 2108}]},
 'augmented': {'brightness': 0.9370387671475437,
  'contrast': 0.8037306995298512,
  'saturation': 0.9335766146012883,
  'hue': 0.0921173427008942,
  'transpose': 0,
  'hflip': 0,
  'vflip': 1}}

In [8]:
# json_data[0]["description"]["type"]
# json_data[0]["description"]["image"]

# "xtl": json_data[0]["annotations"]["points"][0]["xtl"]
# "ytl": json_data[0]["annotations"]["points"][0]["ytl"]
# "xbr": json_data[0]["annotations"]["points"][0]["xbr"]
# "ybr": json_data[0]["annotations"]["points"][0]["ybr"]

In [9]:
# json_data[0]["description"]["width"]
# json_data[0]["description"]["height"]

## Type 0, 1 & Coordinates

In [10]:
# def mk_df(json_data):
#     anno_list = []
#     for i in tqdm(range(len(json_data))):
#         data = {
#         "image": json_data[i]["description"]["image"],
#         "width": json_data[i]["description"]["width"],
#         "height": json_data[i]["description"]["height"],
#         "type": json_data[i]["description"]["type"],
#         "xtl": json_data[i]["annotations"]["points"][0]["xtl"],
#         "ytl": json_data[i]["annotations"]["points"][0]["ytl"],
#         "xbr": json_data[i]["annotations"]["points"][0]["xbr"],
#         "ybr": json_data[i]["annotations"]["points"][0]["ybr"]
#         }
#         anno_list.append(data)
#         df = pd.DataFrame(anno_list)
#     return df

## All data

In [115]:
# def mk_df(json_data):
#     anno_list = []
#     for i in tqdm(range(len(json_data))):
#         data = {
#         "image": json_data[i]["description"]["image"],
#         "date": json_data[i]["description"]["date"],
#         "width": json_data[i]["description"]["width"],
#         "height": json_data[i]["description"]["height"],
        
#         "task": json_data[i]["description"]["task"],
#         "type": json_data[i]["description"]["type"],
#         "region": json_data[i]["description"]["region"],
            
#         "disease": json_data[i]["annotations"]["disease"],
#         "crop": json_data[i]["annotations"]["crop"],
#         "area": json_data[i]["annotations"]["area"],
#         "grow": json_data[i]["annotations"]["grow"],
#         "risk": json_data[i]["annotations"]["risk"],            

#         "xtl": json_data[i]["annotations"]["points"][0]["xtl"],
#         "ytl": json_data[i]["annotations"]["points"][0]["ytl"],
#         "xbr": json_data[i]["annotations"]["points"][0]["xbr"],
#         "ybr": json_data[i]["annotations"]["points"][0]["ybr"]
#         }
#         anno_list.append(data)
#         df = pd.DataFrame(anno_list)
#     return df

## 증강(augmented)

In [11]:
# def mk_aug_df(json_data):
#     anno_list = []
#     for i in tqdm(range(len(json_data))):
#         data = {
#         "image": json_data[i]["description"]["image"],
            
#         "brightness": json_data[i]["augmented"]["brightness"],
#         "contrast": json_data[i]["augmented"]["contrast"],
#         "saturation": json_data[i]["augmented"]["saturation"],
#         "hue": json_data[i]["augmented"]["hue"],
#         "transpose": json_data[i]["augmented"]["transpose"],
#         "hflip": json_data[i]["augmented"]["hflip"],
#         "vflip": json_data[i]["augmented"]["vflip"]
#         }
#         anno_list.append(data)
#         df = pd.DataFrame(anno_list)
#     return df

----

In [12]:
# json_df = mk_df(json_data)

NameError: name 'mk_df' is not defined

In [13]:
json_df = mk_aug_df(json_data)

100%|████████████████████████████████████████████████████████████████████████████| 19030/19030 [03:25<00:00, 92.55it/s]


In [24]:
json_df["transpose"].value_counts()

0    9666
1    9364
Name: transpose, dtype: int64

In [27]:
json_df[json_df["transpose"]==0]

image  brightness  \
0      V006_77_1_18_11_03_13_1_3248b_20201005_57_a000...    0.937039   
1      V006_77_1_18_11_03_13_1_3248b_20201005_57_a000...    0.924047   
5      V006_77_1_18_11_03_13_1_3248b_20201005_57_a000...    1.176505   
14     V006_77_1_18_11_03_13_1_3248b_20201012_60_a000...    0.879488   
16     V006_77_1_18_11_03_13_1_3248b_20201012_60_a000...    1.002616   
...                                                  ...         ...   
19018  V006_77_1_19_11_03_12_3_9357q_20201130_7_a0008...    1.137208   
19019  V006_77_1_19_11_03_12_3_9357q_20201130_7_a0009...    0.860389   
19025  V006_77_1_19_11_03_12_3_9357q_20201130_8_a0005...    0.843890   
19026  V006_77_1_19_11_03_12_3_9357q_20201130_8_a0006...    1.190951   
19027  V006_77_1_19_11_03_12_3_9357q_20201130_8_a0007...    0.997396   

       contrast  saturation       hue  transpose  hflip  vflip  
0      0.803731    0.933577  0.092117          0      0      1  
1      1.077231    0.826694 -0.107073          0      1      0  
5      1.194930    1.184200 -0.074092          0      1      1  
14     1.045699    1.071696 -0.018600          0      0      0  
16     1.088241    0.898347 -0.085353          0      0      1  
...         ...         ...       ...        ...    ...    ...  
19018  1.065431    0.886747  0.040150          0      0      1  
19019  1.164021    1.012902  0.088448          0      0      0  
19025  0.945712    1.173429  0.056812          0      1      0  
19026  0.831703    1.177705  0.035127          0      1      1  
19027  1.186117    1.072132  0.108848          0      1      1  

[9666 rows x 8 columns]

In [28]:
json_df[json_df["transpose"]==1]

image  brightness  \
2      V006_77_1_18_11_03_13_1_3248b_20201005_57_a000...    0.952289   
3      V006_77_1_18_11_03_13_1_3248b_20201005_57_a000...    1.135234   
4      V006_77_1_18_11_03_13_1_3248b_20201005_57_a000...    1.006822   
6      V006_77_1_18_11_03_13_1_3248b_20201005_57_a000...    1.104227   
7      V006_77_1_18_11_03_13_1_3248b_20201005_57_a000...    1.131098   
...                                                  ...         ...   
19022  V006_77_1_19_11_03_12_3_9357q_20201130_8_a0002...    1.056458   
19023  V006_77_1_19_11_03_12_3_9357q_20201130_8_a0003...    0.849163   
19024  V006_77_1_19_11_03_12_3_9357q_20201130_8_a0004...    0.976776   
19028  V006_77_1_19_11_03_12_3_9357q_20201130_8_a0008...    0.912920   
19029  V006_77_1_19_11_03_12_3_9357q_20201130_8_a0009...    1.000667   

       contrast  saturation       hue  transpose  hflip  vflip  
2      1.000891    1.000285 -0.122792          1      0      0  
3      1.156141    1.103838  0.117623          1      0      0  
4      1.195511    1.170775  0.067498          1      0      1  
6      1.143922    0.853635 -0.009060          1      1      0  
7      0.973660    1.055472 -0.135832          1      1      1  
...         ...         ...       ...        ...    ...    ...  
19022  1.143512    1.107301  0.114502          1      1      1  
19023  0.869756    1.084113 -0.061977          1      0      1  
19024  0.901296    1.161819  0.057600          1      0      1  
19028  1.098607    1.135799 -0.070537          1      1      0  
19029  1.121127    1.103439  0.024480          1      1      0  

[9364 rows x 8 columns]

In [119]:
# json_df.to_csv("data/tomato_label/tomato_set/[라벨]11.토마토_9.증강(training).csv", encoding="utf-8-sig", index=False)
# json_df.to_csv("data/tomato_label/tomato_set/[라벨]11.토마토_9.증강(validation).csv", encoding="utf-8-sig", index=False)

# json_df.to_csv("data/tomato_label/tomato_set/[라벨]11.토마토_0.정상(training).csv", encoding="utf-8-sig", index=False)
# json_df.to_csv("data/tomato_label/tomato_set/[라벨]11.토마토_1.질병(training).csv", encoding="utf-8-sig", index=False)

# json_df.to_csv("data/tomato_label/tomato_set/[라벨]11.토마토_0.정상(validation).csv", encoding="utf-8-sig", index=False)
# json_df.to_csv("data/tomato_label/tomato_set/[라벨]11.토마토_1.질병(validation).csv", encoding="utf-8-sig", index=False)

----

### json 데이터 여기부터 실행 !!!!

####  확인 후 .csv 저장 예정

#### DF copy, df 다시 만들지 않도록 주의

In [29]:
yolo_df = json_df.copy()

In [30]:
yolo_df.head(4)

image  width  height  type  \
0  V006_77_1_18_11_03_13_1_3248b_20201005_57_a000...   4032    3024     1   
1  V006_77_1_18_11_03_13_1_3248b_20201005_57_a000...   4032    3024     1   
2  V006_77_1_18_11_03_13_1_3248b_20201005_57_a000...   3024    4032     1   
3  V006_77_1_18_11_03_13_1_3248b_20201005_57_a000...   3024    4032     1   

   xtl  ytl   xbr   ybr  
0  908  712  3104  2108  
1  927  916  3124  2312  
2  916  908  2312  3104  
3  916  908  2312  3104

## YOLO 기준 컬럼 추가

### cx, cy, w, h

- cx = (xtl+xbr)/2
- cy = (ytl+ybr)/2
- w = xbr-xtl
- h = ybr-ytl

In [163]:
yolo_df["cx"] = (yolo_df["xtl"] + yolo_df["xbr"])/2
yolo_df["cy"] = (yolo_df["ytl"] + yolo_df["ybr"])/2
yolo_df["w"] = yolo_df["xbr"] - yolo_df["xtl"]
yolo_df["h"] = yolo_df["ybr"] - yolo_df["ytl"]

In [164]:
yolo_df

image  width  height  type  \
0     V006_77_1_18_11_03_13_1_3248b_20201012_63.jpg   4032    3024     1   
1    V006_77_1_18_11_03_13_1_7688w_20201111_215.jpg   4032    3024     1   
2    V006_77_1_18_11_03_13_1_7688w_20201112_225.jpg   4032    3024     1   
3     V006_77_1_18_11_03_13_1_8419w_20200918_14.jpg   4032    1908     1   
4     V006_77_1_18_11_03_13_1_8419w_20200918_15.jpg   4032    1908     1   
..                                              ...    ...     ...   ...   
241  V006_77_1_19_11_03_12_3_7711b_20201125_52.jpeg   4032    3024     1   
242  V006_77_1_19_11_03_12_3_7711b_20201125_57.jpeg   4032    3024     1   
243   V006_77_1_19_11_03_12_3_7711b_20201125_9.jpeg   4032    3024     1   
244  V006_77_1_19_11_03_12_3_7711b_20201210_12.jpeg   4032    3024     1   
245   V006_77_1_19_11_03_12_3_9357q_20201130_9.jpeg   4032    3024     1   

      xtl   ytl   xbr   ybr      cx      cy     w     h  
0     844   876  2944  2532  1894.0  1704.0  2100  1656  
1     768   224  3508  2124  2138.0  1174.0  2740  1900  
2    1228   792  3068  1968  2148.0  1380.0  1840  1176  
3     834     0  3162  1713  1998.0   856.5  2328  1713  
4     888     0  3671  1908  2279.5   954.0  2783  1908  
..    ...   ...   ...   ...     ...     ...   ...   ...  
241  1610   674  2202  1230  1906.0   952.0   592   556  
242  1947   636  2375  1168  2161.0   902.0   428   532  
243  1357  1228  2513  2268  1935.0  1748.0  1156  1040  
244   344   380  2520  2612  1432.0  1496.0  2176  2232  
245   545   698  3423  2686  1984.0  1692.0  2878  1988  

[246 rows x 12 columns]

## BB 좌표 비정상 조건 확인

- 이미지의 경우 좌측 상단이 0, 0 => DataFrame 행열 형태와 같음

- xbr, ybr은 0이 될 수 없음
- BOTTOM RIGHT 값이 TOP LEFT 보다 작아서는 안된다.

- 따라서 xbr > xtl , ybr > ytl 의 조건을 만족해야 한다.

## 0이 있으면 안되는 경우

In [165]:
yolo_df[yolo_df.xbr == 0]

Empty DataFrame
Columns: [image, width, height, type, xtl, ytl, xbr, ybr, cx, cy, w, h]
Index: []

In [166]:
yolo_df[yolo_df.ybr == 0]

Empty DataFrame
Columns: [image, width, height, type, xtl, ytl, xbr, ybr, cx, cy, w, h]
Index: []

## 0 이 있어도 괜찮은 경우(그냥 확인)

In [167]:
yolo_df[yolo_df.xtl == 0]

image  width  height  type  xtl  \
73   V006_77_1_18_11_03_13_2_8419w_20201020_269.JPG   4032    3024     1    0   
125  V006_77_1_19_11_03_12_1_4521q_20201026_12.jpeg   3024    3024     1    0   
174  V006_77_1_19_11_03_12_2_4521q_20201119_17.jpeg   3024    3024     1    0   
179  V006_77_1_19_11_03_12_2_4521q_20201119_65.jpeg   3024    3024     1    0   
233  V006_77_1_19_11_03_12_3_4521q_20201119_15.jpeg   3024    3024     1    0   

     ytl   xbr   ybr      cx      cy     w     h  
73   418  4032  2822  2016.0  1620.0  4032  2404  
125  508  3000  2108  1500.0  1308.0  3000  1600  
174   76  2424  2160  1212.0  1118.0  2424  2084  
179    4  3024  2768  1512.0  1386.0  3024  2764  
233  776  2984  2344  1492.0  1560.0  2984  1568

In [168]:
yolo_df[yolo_df.ytl == 0]

image  width  height  type  \
3     V006_77_1_18_11_03_13_1_8419w_20200918_14.jpg   4032    1908     1   
4     V006_77_1_18_11_03_13_1_8419w_20200918_15.jpg   4032    1908     1   
68    V006_77_1_18_11_03_13_2_8419w_20200920_27.jpg   4032    1908     1   
104   V006_77_1_18_11_03_13_3_8419w_20200924_74.jpg   4032    1908     1   
171  V006_77_1_19_11_03_12_2_4521q_20201026_17.jpeg   3024    3024     1   
173   V006_77_1_19_11_03_12_2_4521q_20201026_3.jpeg   3024    3024     1   
175  V006_77_1_19_11_03_12_2_4521q_20201119_29.jpeg   3024    3024     1   
178  V006_77_1_19_11_03_12_2_4521q_20201119_51.jpeg   3024    3024     1   
187  V006_77_1_19_11_03_12_2_5952q_20201026_16.jpeg   4032    3024     1   

      xtl  ytl   xbr   ybr      cx      cy     w     h  
3     834    0  3162  1713  1998.0   856.5  2328  1713  
4     888    0  3671  1908  2279.5   954.0  2783  1908  
68    914    0  3202  1299  2058.0   649.5  2288  1299  
104   992    0  2882  1908  1937.0   954.0  1890  1908  
171    11    0  2153  2452  1082.0  1226.0  2142  2452  
173  1516    0  2372  2140  1944.0  1070.0   856  2140  
175   511    0  2261  3024  1386.0  1512.0  1750  3024  
178   820    0  2544  3020  1682.0  1510.0  1724  3020  
187   796    0  2392  1300  1594.0   650.0  1596  1300

----

##  조건 확인
- 아닌 경우

### xbr > xtl

### ybr > ytl 

In [169]:
yolo_df[yolo_df.xbr < yolo_df.xtl]

Empty DataFrame
Columns: [image, width, height, type, xtl, ytl, xbr, ybr, cx, cy, w, h]
Index: []

In [170]:
yolo_df[yolo_df.xbr < yolo_df.xtl]

Empty DataFrame
Columns: [image, width, height, type, xtl, ytl, xbr, ybr, cx, cy, w, h]
Index: []

----

## cx, cy, w, h 이미지 크기 기준, 비율로 정규화

### 0~1

In [171]:
yolo_df["cx"] = yolo_df["cx"]/yolo_df["width"]
yolo_df["cy"] = yolo_df["cy"]/yolo_df["height"]
yolo_df["w"] = yolo_df["w"]/yolo_df["width"]
yolo_df["h"] = yolo_df["h"]/yolo_df["height"]

In [172]:
yolo_df

image  width  height  type  \
0     V006_77_1_18_11_03_13_1_3248b_20201012_63.jpg   4032    3024     1   
1    V006_77_1_18_11_03_13_1_7688w_20201111_215.jpg   4032    3024     1   
2    V006_77_1_18_11_03_13_1_7688w_20201112_225.jpg   4032    3024     1   
3     V006_77_1_18_11_03_13_1_8419w_20200918_14.jpg   4032    1908     1   
4     V006_77_1_18_11_03_13_1_8419w_20200918_15.jpg   4032    1908     1   
..                                              ...    ...     ...   ...   
241  V006_77_1_19_11_03_12_3_7711b_20201125_52.jpeg   4032    3024     1   
242  V006_77_1_19_11_03_12_3_7711b_20201125_57.jpeg   4032    3024     1   
243   V006_77_1_19_11_03_12_3_7711b_20201125_9.jpeg   4032    3024     1   
244  V006_77_1_19_11_03_12_3_7711b_20201210_12.jpeg   4032    3024     1   
245   V006_77_1_19_11_03_12_3_9357q_20201130_9.jpeg   4032    3024     1   

      xtl   ytl   xbr   ybr        cx        cy         w         h  
0     844   876  2944  2532  0.469742  0.563492  0.520833  0.547619  
1     768   224  3508  2124  0.530258  0.388228  0.679563  0.628307  
2    1228   792  3068  1968  0.532738  0.456349  0.456349  0.388889  
3     834     0  3162  1713  0.495536  0.448899  0.577381  0.897799  
4     888     0  3671  1908  0.565352  0.500000  0.690228  1.000000  
..    ...   ...   ...   ...       ...       ...       ...       ...  
241  1610   674  2202  1230  0.472718  0.314815  0.146825  0.183862  
242  1947   636  2375  1168  0.535962  0.298280  0.106151  0.175926  
243  1357  1228  2513  2268  0.479911  0.578042  0.286706  0.343915  
244   344   380  2520  2612  0.355159  0.494709  0.539683  0.738095  
245   545   698  3423  2686  0.492063  0.559524  0.713790  0.657407  

[246 rows x 12 columns]

## cx cy w h, 0~1 사이 값인지 확인

- 0~1 이 아닌 경우를 확인하면 되지만 코드 연습 :)
    - 혹시나 분석에 사용될 수 있음(conf 수치와 비교,, 평균bb사이즈별 학습 차이, 모델별로.. 등등)

In [173]:
yolo_df[(yolo_df["cx"] >= 0) & (yolo_df["cx"] <= 1)].cx

0      0.469742
1      0.530258
2      0.532738
3      0.495536
4      0.565352
         ...   
241    0.472718
242    0.535962
243    0.479911
244    0.355159
245    0.492063
Name: cx, Length: 246, dtype: float64

In [174]:
print("(df의 row수) - (조건에 맞는 row수)\n =>", len(yolo_df) - len(yolo_df[(yolo_df["cx"] >= 0) & (yolo_df["cx"] <= 1)].cx))
print("(df의 row수) - (조건에 맞는 row수)\n =>", len(yolo_df) - len(yolo_df[(yolo_df["cy"] >= 0) & (yolo_df["cy"] <= 1)].cy))
print("(df의 row수) - (조건에 맞는 row수)\n =>", len(yolo_df) - len(yolo_df[(yolo_df["w"] >= 0) & (yolo_df["w"] <= 1)].w))
print("(df의 row수) - (조건에 맞는 row수)\n =>", len(yolo_df) - len(yolo_df[(yolo_df["h"] >= 0) & (yolo_df["h"] <= 1)].h))

(df의 row수) - (조건에 맞는 row수)
 => 0
(df의 row수) - (조건에 맞는 row수)
 => 0
(df의 row수) - (조건에 맞는 row수)
 => 0
(df의 row수) - (조건에 맞는 row수)
 => 0


## 0~1 아닌 것 조회

In [175]:
# 0보다 작은 경우가... 설마 있을까?
print(yolo_df[(yolo_df["cx"] < 0)].cx)
print(yolo_df[(yolo_df["cy"] < 0)].cy)
print(yolo_df[(yolo_df["w"] < 0)].w)
print(yolo_df[(yolo_df["h"] < 0)].h)

Series([], Name: cx, dtype: float64)
Series([], Name: cy, dtype: float64)
Series([], Name: w, dtype: float64)
Series([], Name: h, dtype: float64)


In [176]:
print(yolo_df[(yolo_df["cx"] > 1)])
print(yolo_df[(yolo_df["cy"] > 1)])
print(yolo_df[(yolo_df["w"] > 1)])
print(yolo_df[(yolo_df["h"] > 1)])

Empty DataFrame
Columns: [image, width, height, type, xtl, ytl, xbr, ybr, cx, cy, w, h]
Index: []
Empty DataFrame
Columns: [image, width, height, type, xtl, ytl, xbr, ybr, cx, cy, w, h]
Index: []
Empty DataFrame
Columns: [image, width, height, type, xtl, ytl, xbr, ybr, cx, cy, w, h]
Index: []
Empty DataFrame
Columns: [image, width, height, type, xtl, ytl, xbr, ybr, cx, cy, w, h]
Index: []


In [177]:
# yolo_df[(yolo_df["cx"] > 1)].image

----

In [178]:
yolo_df.loc[0, "image"]

'V006_77_1_18_11_03_13_1_3248b_20201012_63.jpg'

# DataFrame.reindex(columns=["a", "b"])

## 컬럼 정리, copy

In [179]:
yolo_df.head(0)

Empty DataFrame
Columns: [image, width, height, type, xtl, ytl, xbr, ybr, cx, cy, w, h]
Index: []

In [180]:
yolo_data = yolo_df.copy()
yolo_data = yolo_data.reindex(columns=["type", "cx", "cy", "w", "h"])

In [181]:
yolo_data.head(0)

Empty DataFrame
Columns: [type, cx, cy, w, h]
Index: []

## 타입 => 문자열로 수정

In [182]:
yolo_data = yolo_data.astype("str")

In [183]:
yolo_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 246 entries, 0 to 245
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   type    246 non-null    object
 1   cx      246 non-null    object
 2   cy      246 non-null    object
 3   w       246 non-null    object
 4   h       246 non-null    object
dtypes: object(5)
memory usage: 9.7+ KB


## .json -> .txt (YOLO)

In [184]:
# 개수 한번 더 확인
len(yolo_df.image)

246

## 파일 저장 경로!!!! 확인하고 실행!!!!

In [185]:
num_rows = len(yolo_data)

for i in tqdm(range(num_rows)):
# for i in range(4):
    value = " ".join(yolo_data.iloc[i])

    # 파일 저장 위치!!
    filename = f"data/tomato_yolo/val/labels/{yolo_df['image'][i].split('.')[0]}.txt"
    
    with open(filename, "w") as f:
        f.write(value)
        

num_files = len(os.listdir("data/tomato_yolo/val/labels"))
print(f"경로 내 총 파일 수 : {num_files}")

100%|██████████████████████████████████████████████████████████████████████████████| 246/246 [00:00<00:00, 1665.61it/s]

경로 내 총 파일 수 : 4669


In [180]:
# for i in range(len(yolo_data)):
#     print(os.listdir("C:/juna27/PythonWork/AI/data/tomato_yolo/val/labels/")[i].split(".")[1], end=",")

train 정상
13509
train 질병
1967, 15476



val 정상
4423
val 질병
246, 4669



## 메모

lap

1

Train 정상 txt

경로 내 총 파일 수 : 13509

In [222]:
13509 + 1967

15476

2

Train 질병 txt

경로 내 총 파일 수 : 15476

----

3

Val 정상 txt

경로 내 총 파일 수 : 4423

In [298]:
4423 + 246

4669

4

Val 질병 txt

경로 내 총 파일 수 : 4669

In [301]:
from datetime import datetime
datetime.now()

datetime.datetime(2023, 6, 24, 20, 8, 45, 976693)

- 1차 시설 작물 질병 진단 (토마토)
    - 전처리 완료
    - 시간 체크 : datetime.datetime(2023, 6, 24, 20, 8, 45, 976693)

### TODO

- type 0, 1로 Detection 이상 없음을 확인 - 완료

- 어떤 class를 선택 할 것인지 결정하고 전처리 하기

In [307]:
# C:/juna27/PythonWork/AI/data/tomato_yolo

## YOLOv8

C:\Users\acorn\AppData\Roaming\Ultralytics



settings.yaml 확인

In [303]:
import ultralytics

In [308]:
ultralytics.checks()

Ultralytics YOLOv8.0.121  Python-3.9.13 torch-2.0.1+cpu CPU
Setup complete  (8 CPUs, 7.9 GB RAM, 192.2/222.4 GB disk)


### yolov8
- detection models
    - n
    - s
    - m
    - l
    - x

In [305]:
from ultralytics import YOLO

model = YOLO("yolov8n.pt")
# model = YOLO("yolov8s.pt")

In [432]:
# print(len(model.names))
# print("\n")
# print(model.names)

In [306]:
import yaml

data = {
    "train" : "train/images/",
    "val" : "val/images/",
#     "test" : "tomato_yolo/test/images/",
    "nc" : 2,
    "names": ['healthy', 'disease']
}

with open("data/tomato_yolo/tom_out.yaml", "w") as f:
    yaml.dump(data, f)

In [309]:
start_time = datetime.now()
print("Start Training :", start_time)
##########


model.train(data="C:/juna27/PythonWork/AI/data/tomato_yolo/tom_out.yaml",
            epochs=1, patience=20,
            batch=32, imgsz=416)


##########
end_time = datetime.now()
print("\nTime taken :", end_time-start_time)

Ultralytics YOLOv8.0.121  Python-3.9.13 torch-2.0.1+cpu CPU
yolo\engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=C:/juna27/PythonWork/AI/data/tomato_yolo/tom_out.yaml, epochs=1, patience=20, batch=32, imgsz=416, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=None, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=0, resume=False, amp=True, fraction=1.0, profile=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, show=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, vid_stride=1, line_width=None, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=F

## runs/detect 경로는 ipynb 경로와 같음

In [311]:
# results = model.predict(source="tomato_yolo/test/images/", save=True)

C:\Users\acorn\AppData\Roaming\Ultralytics



settings.yaml 확인